tests/common.go

```diff
@@ -186,6 +199,157 @@ func InitInstance() {
 		err = Inst().Backup.Init(Inst().S.String(), Inst().N.String(), Inst().V.String(), token)
 		expect(err).NotTo(haveOccurred())
 	}
+	if Inst().TestRailMilestoneName != "" {
+		client = testrail.NewClient(Inst().TestRailHostname, Inst().TestRailUsername, Inst().TestRailPassword, true)
+		CreateMilestoneIfNotExists()
+	} else {
+		logrus.Infof("No milestone provided, skipping updates to testrail")
+	}
+}
+
+func GetMilestoneByProjectId(projectId int) int {
+	logrus.Infof("Getting the milestone ID")
+	milestones, err := client.GetMilestones(projectId)
+	if err != nil {
+		logrus.Infof("Error in getting milestone: %s", err)
+		return 0
+	}
+	for _, ms := range milestones {
+		if ms.Name == Inst().TestRailMilestoneName {
+			return ms.ID
+		}
+	}
+	return 0
+}
+
+func GetRunId(projectId int, filter testrail.RequestFilterForRun) int {
+	allRuns, err := client.GetRuns(projectId, filter)
+	if err != nil {
+		processError(fmt.Errorf("Error in getting all jobs from milestone %s ", err))
+	}
+	for _, run := range allRuns {
+		if run.Name == Inst().JenkinsJobName {
+			logrus.Infof("Run %s already exsits in milestone not adding it\n", Inst().JenkinsJobName)
+			return run.ID
+		}
+	}
+	return 0
+}
+
+func GetTestIdForRunId(run_id int, testRailId int) int {
+	logrus.Infof("Getting the Tests Id for run: %d", run_id)
+	tests, err := client.GetTests(run_id)
+	if err != nil {
+		processError(fmt.Errorf("Error getting tests for existsing run %s", run_id))
+	}
+	for _, test := range tests {
+		if test.CaseID == testRailId {
+			logrus.Infof("Test already exists in the run, not adding new test")
+			return test.CaseID
+		}
+	}
+	return 0
+
+}
+
+func CreateMilestoneIfNotExists() {
+	logrus.Infof("Connection to testrail successful")
+	milestoneId = GetMilestoneByProjectId(pwxProjectId)
+	if milestoneId == 0 {
+		logrus.Infof("Creating milesone, since it does not exists")
+		sendatbleMs := testrail.SendableMilestone{
+			Name:        Inst().TestRailMilestoneName,
+			Description: "Created from Automation",
+		}
+		msCreated, err := client.AddMilestone(pwxProjectId, sendatbleMs)
+		if err != nil {
+			processError(fmt.Errorf("error in creating milestone: %s", Inst().TestRailMilestoneName))
```

Comment:
wrong usage of processError function, it takes `err error, errChan ...*chan error` and I think you don't want to fail the test because milestone already exists.